_i18n/en/_posts/2021-09-03-frida-15-1-released.markdown

```diff
@@ -0,0 +1,128 @@
+[include/swift/ABI/Metadata.h][] at the time of writing. This data structure
+includes the type name, its fields, its methods (if applicable,) and other useful
+data depending on the type at hand. These data structures are pointed to by
+relative pointers (defined in [include/swift/Basic/RelativePointer.h][].) In
+Mach-O binaries, these are stored in the `__swift5_types` section.
+
+So to dump types, Frida basically iterates over these data structures and
+parses them along the way, very similar to what [dsdump][] does, except that you
+don't have to build the Swift compiler to in order to tinker with it.
+
+Frida also has the advantage of being able to probe into
+internal Apple dylibs written in Swift, and that's because we don't need to
+parse the `dyld_shared_cache` thanks to the private `getsectiondata` API, whch
+gives us section offsets hassle-free.
+
+Once we have the metadata, we're able to easily create JavaScript wrappers for
+object instances and values of different types.
+
+## Conventions
+
+To be on par with the Objective-C bridge, the Swift bridge has to support
+calling Swift functions, which also proved to be not as straight forward.
+
+Swift defines its own calling convention, `swiftcall`, which, to put it
+succinctly, tries to be as efficient as possible. That means, not wasting load
+and store instructions on structs that are smaller than 4 registers-worth of
+size. That is, to pass those kinds of structs directly in registers. And since
+that could quickly over-book our precious 8 argument registers
+(on AARCH64 `x0`-`x7`), it doesn't use the first register for the `self`
+argument. It also defines an `error` register where callees can store errors
+which they throw.
+
+What we just described above is termed "physical lowering" in the Swift compiler
+docs, and it's implemented by the back-end, LLVM.
+
+The process that precedes physical lowering is termed "semantic lowering," which
+is the compiler front-end figuring out who "owns" a value and whether
+it's direct or indirect. Some structs, even though they might be smaller than
+4 registers, have to be passed indirectly, because, for example, they are
+generic and thus their exact memory layout is not known at compile time, or
+because they include a weak reference that has to be in-memory at all times.
+
+Frida had to implement both semantic and physical lowering in order to be able
+to call Swift functions. Physical lowering is implemented using JIT-compiled
+adapter functions (thanks to the `Arm64Writer` API) that does the necessary
+`SystemV`-`swiftcall` translation. Semantic lowering is implemented by utilizing
+the type's metadata to figure out whether we should pass a value directly or
+not.
+
+The compiler [docs][] are a great resource to learn more about the calling
+convention.
+
+## Interception
+
+Because Swift passes structs directly in registers, there isn't a 1-to-1 mapping
+between registers and actual arguments, as is the case for SystemV.
```

Comment:
```suggestion
between registers and actual arguments.
```

Structs passed by value might end up sliced across registers with SystemV as well – so maybe it's better to just drop that last part?

Reply:
Ah, interesting. Definitely yeah, let's drop it.